coqpp/coqpp_main.ml

```diff
@@ -357,22 +357,32 @@ let print_atts_right fmt = function
     let nota = match atts with [_] -> "" | _ -> "Attributes.Notations." in
     fprintf fmt "(Attributes.parse %s%a atts)" nota aux atts
 
-let print_body_wrapper fmt r =
-  match r.vernac_state with
-  | Some "proof" ->
-    fprintf fmt "let proof = (%a) ~pstate:st.Vernacstate.proof in { st with Vernacstate.proof }" print_code r.vernac_body
-  | None ->
-    fprintf fmt "let () = %a in st" print_code r.vernac_body
-  | Some x ->
-    fatal ("unsupported state specifier: " ^ x)
-
-let print_body_fun fmt r =
-  fprintf fmt "let coqpp_body %a%a ~st = @[%a@] in "
-    print_binders r.vernac_toks print_atts_left r.vernac_atts print_body_wrapper r
-
-let print_body fmt r =
-  fprintf fmt "@[(%afun %a~atts@ ~st@ -> coqpp_body %a%a ~st)@]"
-    print_body_fun r print_binders r.vernac_toks
+let understand_state = function
+  | "close_proof" -> "VtCloseProof", false
+  | "maybe_open_proof" -> "VtMaybeOpenProof", true
+  | "open_proof" -> "VtOpenProof", true
+  | "proof" -> "VtModifyProof", false
+  | "proof_opt_query" -> "VtReadProofOpt", false
+  | "proof_query" -> "VtReadProof", false
+  | s -> fatal ("unsupported state specifier: " ^ s)
+
+let print_body_state state fmt r =
+  let state = match r.vernac_state with Some _ as s -> s | None -> state in
+  match state with
+  | None -> fprintf fmt "Vernacextend.VtDefault (fun () -> %a)" print_code r.vernac_body
+  | Some "CUSTOM" -> print_code fmt r.vernac_body
```

Comment:
@SkySkimmer what do you think of removing CUSTOM, IMHO it defeats the purpose of this PR and leaving it there is just opening the door for people keep doing hacks at this level.

Reply:
I think it's fine to be able to have multiple possible behaviours for 1 parsing rule. Earlier we had some discussion https://github.com/coq/coq/pull/10050#issuecomment-491909057
> >    You can also make a different block of grammar entries.
>
> Well that would imply a large stylistic change in the rules and I am not sure the people working on the grammar doc would like that a lot.

In the same way I don't see why we should have to split a parsing rul to be able to get multiple proof-related behaviours.

Reply:
OK then.